File: src/audit_reports/document_related.py
"""Preserve other PDFs in a registered report's archive as separate documents."""
from __future__ import annotations

import hashlib
import io
import json
import zipfile

from .document_acquisition import unwrap_pdf
from .document_corpus import Filing
from .document_corpus_store import CorpusStore, PREFIX


def _sha(body):
    return hashlib.sha256(body).hexdigest()


def _verified(store, entry, expected_key):
    if entry['key'] != expected_key:
        raise ValueError('Related source reference is outside its expected namespace')
    body, _ = store._read(expected_key)
    if body is None or len(body) != entry['bytes'] or _sha(body) != entry['sha256']:
        raise ValueError('Related source evidence bytes differ from their receipt')
    return body


def related_sources(store: CorpusStore, filing: Filing) -> tuple[dict, list[tuple[dict, bytes]]]:
    """Verify a retained origin/ZIP and enumerate every non-primary PDF member."""
    base = f'{PREFIX}origins/{filing.bank_ticker}/{filing.period}/{filing.kind}/'
    body, _ = store._read(base + 'index.json')
    if body is None:
        raise ValueError('No retained official-origin comparison for this filing')
    index = json.loads(body)
    if (index.get('schema_version') != 'document-origin-index-1' or index.get('filing') != filing.as_dict()
            or index.get('semantically_verified') is not False or index.get('current') not in index.get('revisions', [])):
        raise ValueError('Related source origin index has an invalid filing binding')
    current = index['current']
    receipt = json.loads(_verified(store, current, base + current['sha256'] + '.json'))
    if (receipt.get('schema_version') != 'document-origin-review-1' or receipt.get('filing') != filing.as_dict()
            or receipt.get('semantically_verified') is not False
            or receipt.get('status') != current['status'] or receipt.get('checked_at') != current['checked_at']):
        raise ValueError('Related source origin receipt differs from its index')
    if not receipt.get('transport'):
        raise ValueError('The official source response is unavailable')
    entry = receipt['transport']
    transport = _verified(store, entry, f"{PREFIX}transports/{entry['sha256']}/original.bin")
    if not transport.startswith(b'PK\x03\x04'):
        if receipt.get('selection', {}).get('unselected_pdf_members'):
            raise ValueError('Related PDF members claimed without an archive')
        return receipt, []
    selection = receipt.get('selection', {})
    primary_name = selection.get('archive_member')
    with zipfile.ZipFile(io.BytesIO(transport)) as archive:
        members = [m for m in archive.infolist() if not m.is_dir()]
        if len({m.filename for m in members}) != len(members):
            raise ValueError('Duplicate archive member names require source review')
        actual = [{'name': m.filename, 'bytes': m.file_size, 'sha256': _sha(archive.read(m))} for m in members]
        if actual != selection.get('archive_members'):
            raise ValueError('Retained archive member inventory differs from source bytes')
        if primary_name not in {m.filename for m in members} or not receipt.get('origin_pdf'):
            raise ValueError('Archive has no verified primary report selection')
        primary, _ = unwrap_pdf(archive.read(primary_name))
        if _sha(primary) != receipt['origin_pdf']['sha256'] or len(primary) != receipt['origin_pdf']['bytes']:
            raise ValueError('Archive primary member differs from its retained report')
        others = [m for m in actual if m['name'].lower().endswith('.pdf') and m['name'] != primary_name]
        if others != selection.get('unselected_pdf_members', []):
            raise ValueError('A related PDF member was omitted or invented')
        sources = []
        for member in others:
            relation = {'schema_version': 'related-source-binding-1', 'filing': filing.as_dict(),
                        'relationship': 'other_pdf_in_same_registered_source_archive',
                        'transport_sha256': entry['sha256'], 'transport_key': entry['key'],
                        'primary_pdf_sha256': receipt['origin_pdf']['sha256'], 'primary_member_name': primary_name,
                        'member': member,
                        'semantically_verified': False}
            sources.append((relation, archive.read(member['name'])))
    return receipt, sources


class RelatedCorpusStore(CorpusStore):
    """Reuse native preservation with a separate archive-member index namespace."""
    def __init__(self, store: CorpusStore, relation: dict):
        super().__init__(store.client, store.bucket)
        self.relation = relation
        self.filing = Filing(**relation['filing'])
        if (relation.get('schema_version') != 'related-source-binding-1' or relation.get('semantically_verified') is not False
                or relation.get('relationship') != 'other_pdf_in_same_registered_source_archive'
                or relation['member']['name'] == relation['primary_member_name']):
            raise ValueError('Invalid related document binding')
        for value in (relation['transport_sha256'], relation['primary_pdf_sha256'], relation['member']['sha256']):
            if not isinstance(value, str) or len(value) != 64 or any(c not in '0123456789abcdef' for c in value):
                raise ValueError('Invalid related source digest')

    def index_key(self, filing: Filing) -> str:
        if filing != self.filing:
            raise ValueError('Related document store cannot address another filing')
        return (f'{PREFIX}related/{filing.bank_ticker}/{filing.period}/{filing.kind}/'
                f"{self.relation['transport_sha256']}/{self.relation['member']['sha256']}.json")

    def _update_index(self, filing, update):
        def bound(index):
            if 'relationship' in index and index['relationship'] != self.relation:
                raise ValueError('Related document index has a different archive binding')
            index['relationship'] = self.relation
            update(index)
        return super()._update_index(filing, bound)

    def publish(self, records, original, evidence):
        # A caller-supplied relationship is not sufficient proof: locate the
        # exact member again in the retained transport before any publication.
        entry = {'key': self.relation['transport_key'], 'sha256': self.relation['transport_sha256']}
        body, _ = self._read(entry['key'])
        if entry['key'] != f"{PREFIX}transports/{entry['sha256']}/original.bin" or body is None or _sha(body) != entry['sha256']:
            raise ValueError('Related document transport binding changed')
        with zipfile.ZipFile(io.BytesIO(body)) as archive:
            primary = [m for m in archive.infolist() if m.filename == self.relation['primary_member_name']]
            if len(primary) != 1 or _sha(unwrap_pdf(archive.read(primary[0]))[0]) != self.relation['primary_pdf_sha256']:
                raise ValueError('Related document primary report binding changed')
            matching = [m for m in archive.infolist() if m.filename == self.relation['member']['name']]
            if len(matching) != 1:
                raise ValueError('Related document member binding is ambiguous')
            raw = archive.read(matching[0])
            if _sha(raw) != self.relation['member']['sha256'] or len(raw) != self.relation['member']['bytes']:
                raise ValueError('Related document member bytes changed')
            canonical, _ = unwrap_pdf(raw)
            if canonical != original.read_bytes():
                raise ValueError('Related document original differs from its archive member')
        return super().publish(records, original, evidence)

File: web/app/lib/document-related.ts
/** Resolve an attachment only through its retained official archive relationship. */
import { CORPUS_PREFIX, parseCorpusRevision, type CorpusBucket, type FilingIdentity } from "./document-corpus";
import { getOriginReview } from "./document-origin";

const record = (v: unknown): v is Record<string, unknown> => typeof v === "object" && v !== null && !Array.isArray(v);

export async function getRelatedRevision(bucket: CorpusBucket, filing: FilingIdentity, memberHash: string) {
  if (!/^[a-f0-9]{64}$/.test(memberHash)) throw new Error("Invalid related document hash");
  const origin = await getOriginReview(bucket, filing);
  const members = origin?.selection?.unselected_pdf_members?.filter(m => m.sha256 === memberHash) ?? [];
  if (!origin?.transport || !origin.origin_pdf || members.length !== 1) {
    throw new Error("Related document is absent or ambiguous in the verified source archive");
  }
  const member = members[0];
  const key = `${CORPUS_PREFIX}related/${filing.bank_ticker}/${filing.period}/${filing.kind}/${origin.transport.sha256}/${memberHash}.json`;
  const object = await bucket.get(key);
  if (!object) return null;
  if (object.size > 8_000_000) throw new Error("Oversized related document index");
  const index: unknown = await object.json();
  const relation = record(index) ? index.relationship : null;
  if (!record(relation) || relation.schema_version !== "related-source-binding-1"
      || relation.relationship !== "other_pdf_in_same_registered_source_archive" || relation.semantically_verified !== false
      || !record(relation.filing) || Object.entries(filing).some(([k, v]) => (relation.filing as Record<string, unknown>)[k] !== v)
      || relation.transport_sha256 !== origin.transport.sha256 || relation.transport_key !== origin.transport.key
      || relation.primary_pdf_sha256 !== origin.origin_pdf.sha256 || relation.primary_member_name !== origin.selection?.archive_member
      || !record(relation.member) || Object.entries(member).some(([k, v]) => (relation.member as Record<string, unknown>)[k] !== v)) {
    throw new Error("Related document index differs from its official source archive");
  }
  const revision = parseCorpusRevision(index, filing);
  // For ordinary PDF members, the origin's independently retained member hash
  // binds the current source directly. A wrapped member needs a separate
  // verified wrapper binding; never accept another valid filing revision here.
  if (revision && revision.source.pdf_sha256 !== memberHash) {
    throw new Error("Related PDF bytes cannot be matched directly to the retained archive member");
  }
  if (revision && (!record(index) || !record(index.current) || index.current.semantic_verification !== "not_performed")) {
    throw new Error("Related source claims unsupported semantic approval");
  }
  return revision;
}
